File: android/app/src/main/java/com/vocahq/vocaphone/local/SherpaRecognizer.kt
package com.vocahq.vocaphone.local

import com.k2fsa.sherpa.onnx.OfflineCanaryModelConfig
import com.k2fsa.sherpa.onnx.OfflineDolphinModelConfig
import com.k2fsa.sherpa.onnx.FeatureConfig
import com.k2fsa.sherpa.onnx.OfflineModelConfig
import com.k2fsa.sherpa.onnx.OfflineMoonshineModelConfig
import com.k2fsa.sherpa.onnx.OfflineNemoEncDecCtcModelConfig
import com.k2fsa.sherpa.onnx.OfflineParaformerModelConfig
import com.k2fsa.sherpa.onnx.OfflineRecognizer
import com.k2fsa.sherpa.onnx.OfflineRecognizerConfig
import com.k2fsa.sherpa.onnx.OfflineSenseVoiceModelConfig
import com.k2fsa.sherpa.onnx.OfflineTransducerModelConfig
import com.vocahq.vocaphone.core.TranscriptionQuality
import java.io.File

/**
 * One sherpa-onnx recognizer.
 *
 * Every family in [SherpaFamily] differs only in which pinned files go into
 * which slot of `OfflineModelConfig`, so adding a family is a `when` branch here
 * plus an entry in [SherpaModelCatalog].
 */
internal class SherpaRecognizer private constructor(
    private val recognizer: OfflineRecognizer,
    /**
     * Whether this recognizer was built to translate, which changes how a
     * recording too long for one decode may be cut up. See [transcribe].
     */
    private val translating: Boolean = false,
) {
    /**
     * Decodes the whole recording, in windows if it is longer than one decode
     * should be.
     *
     * Those windows are where translation and transcription part company. Each
     * window retains a little of the one before it so that a word on the
     * boundary survives, and the merger matches the repeated words and removes
     * them. That second half only works if the same audio returns the same
     * words, which a transcriber promises and a translator does not: the
     * retained audio is translated again inside a different sentence and comes
     * back worded differently, so there is nothing to pair.
     *
     * Left on, the merger would then be matching text it was never able to
     * align, and the only thing it can still find is a phrase the speaker
     * genuinely said twice — which it would delete. A translated seam is
     * therefore left exactly as it decoded. The audio overlap stays: almost all
     * of it is the measured quiet run the boundary was found in, which
     * duplicates nothing, and where the boundary was misjudged a word repeated
     * beats a word lost.
     */
    fun transcribe(samples: FloatArray): SherpaTranscript {
        var transcript = SherpaTranscript.EMPTY
        var previousEnd = 0
        var loudestSoFar = 0.0
        SherpaLongAudio.chunks(samples).forEach { chunk ->
            // Whether a short empty answer is ordinary is a question about the
            // audio this window did not inherit from the one before it. A final
            // window that is only the retained overlap has nothing new in it,
            // and an empty answer there is correct. A final window carrying a
            // whole further sentence is the reported failure, and it is that
            // whether or not an earlier window already produced text -- judging
            // it by the transcript so far is what let a closing sentence
            // disappear behind a successful opening one.
            val newRegion = SherpaLongAudio.newRegion(samples, chunk, previousEnd)
            val newRegionLevel = SherpaLongAudio.loudestFrame(newRegion)
            val carriesNewSpeech = SherpaLongAudio.carriesRecoverableSpeech(
                newRegion = newRegion,
                inheritsAudio = chunk.overlapsPrevious,
                loudestFrame = newRegionLevel,
                loudestFrameSoFar = loudestSoFar,
            )
            val chunkResult = SherpaEmptyChunkRecovery.decode(
                samples = samples.copyOfRange(chunk.start, chunk.endExclusive),
                decodeOnce = ::decode,
                deduplicateOverlap = !translating,
                recoverAudibleShortInput = carriesNewSpeech,
            )
            loudestSoFar = maxOf(loudestSoFar, newRegionLevel)
            previousEnd = chunk.endExclusive
            transcript = transcript.append(
                chunkResult,
                deduplicateOverlap = chunk.overlapsPrevious && !translating,
            )
        }
        return transcript.copy(text = transcript.text.trim())
    }

    /** Decodes one bounded window, recovering an empty result with smaller windows. */
    fun transcribeChunk(samples: FloatArray): SherpaTranscript = SherpaEmptyChunkRecovery.decode(
        samples = samples,
        decodeOnce = ::decode,
        deduplicateOverlap = !translating,
    )

    private fun decode(samples: FloatArray): SherpaTranscript {
        val stream = recognizer.createStream()
        return try {
            stream.acceptWaveform(samples, SherpaLongAudio.SAMPLE_RATE)
            recognizer.decode(stream)
            val result = recognizer.getResult(stream)
            SherpaTranscript(
                text = result.text.trim(),
                language = SherpaTranscript.languageCode(result.lang),
            )
        } finally {
            stream.release()
        }
    }

    fun release() = recognizer.release()

    companion object {
        /**
         * @param language a two-letter code, or "auto" to let the model decide.
         *   Only the families that accept a language hint use it.
         * @param translateTo the language to translate into, or empty to
         *   transcribe. Canary is the only family that can honour it; see
         *   [com.vocahq.vocaphone.core.ModelTranslationSupport].
         */
        fun create(
            model: LocalModelDescriptor,
            directory: File,
            language: String,
            threads: Int,
            quality: TranscriptionQuality,
            translateTo: String = "",
        ): SherpaRecognizer {
            val family = requireNotNull(model.sherpaFamily) {
                "${model.displayName} has no sherpa-onnx family"
            }
            fun path(name: String): String {
                val file = File(directory, name)
                require(file.isFile) { "${model.displayName} is missing $name" }
                return file.absolutePath
            }

            // The quantized file where the model ships one, the plain file
            // where it does not. Upstream quantizes per graph rather than per
            // model: GigaAM's transducer ships an int8 encoder beside a
            // full-precision decoder and joiner, because those two are small
            // enough that quantizing them costs accuracy for nothing.
            fun quantizedOrPlain(stem: String): String =
                if (File(directory, "$stem.int8.onnx").isFile) path("$stem.int8.onnx")
                else path("$stem.onnx")

            val tokens = path("tokens.txt")
            val modelConfig = when (family) {
                SherpaFamily.NEMO_TRANSDUCER -> OfflineModelConfig(
                    transducer = OfflineTransducerModelConfig(
                        encoder = quantizedOrPlain("encoder"),
                        decoder = quantizedOrPlain("decoder"),
                        joiner = quantizedOrPlain("joiner"),
                    ),
                )

                SherpaFamily.SENSE_VOICE -> OfflineModelConfig(
                    senseVoice = OfflineSenseVoiceModelConfig(
                        model = path("model.int8.onnx"),
                        language = if (language == "auto") "" else language,
                        useInverseTextNormalization = true,
                    ),
                )

                SherpaFamily.MOONSHINE -> OfflineModelConfig(
                    moonshine = OfflineMoonshineModelConfig(
                        preprocessor = path("preprocess.onnx"),
                        encoder = path("encode.int8.onnx"),
                        uncachedDecoder = path("uncached_decode.int8.onnx"),
                        cachedDecoder = path("cached_decode.int8.onnx"),
                    ),
                )

                // The other three fields stay empty, which is how sherpa-onnx
                // tells the two Moonshine layouts apart.
                SherpaFamily.MOONSHINE_V2 -> OfflineModelConfig(
                    moonshine = OfflineMoonshineModelConfig(
                        encoder = path("encoder_model.ort"),
                        mergedDecoder = path("decoder_model_merged.ort"),
                    ),
                )

                SherpaFamily.DOLPHIN_CTC -> OfflineModelConfig(
                    dolphin = OfflineDolphinModelConfig(model = path("model.int8.onnx")),
                )

                // The one family that can translate. Equal source and target is
                // transcription; differing them is what Canary was trained for,
                // and "auto" has to become a real code because the config has no
                // detection mode — English is the safest guess and the one
                // upstream's own examples use.
                SherpaFamily.CANARY -> {
                    val source = if (language == "auto") "en" else language
                    OfflineModelConfig(
                        canary = OfflineCanaryModelConfig(
                            encoder = path("encoder.int8.onnx"),
                            decoder = path("decoder.int8.onnx"),
                            srcLang = source,
                            tgtLang = translateTo.ifEmpty { source },
                            usePnc = true,
                        ),
                    )
                }

                // Upstream is inconsistent about whether the single CTC graph is
                // quantized, so whichever one the catalog pinned is the one here.
                SherpaFamily.NEMO_CTC -> OfflineModelConfig(
                    nemo = OfflineNemoEncDecCtcModelConfig(
                        model = path(model.primaryFile.path),
                    ),
                )

                SherpaFamily.PARAFORMER -> OfflineModelConfig(
                    paraformer = OfflineParaformerModelConfig(model = path("model.int8.onnx")),
                )
            }.copy(
                tokens = tokens,
                numThreads = threads,
                provider = "cpu",
                modelType = family.sherpaModelType,
            )

            // Never `quality.sherpaDecodingMethod` on its own: a family that
            // does not support beam search answers it by killing the process.
            val decodingMethod = family.decodingMethod(quality)
            return SherpaRecognizer(
                OfflineRecognizer(
                    assetManager = null,
                    config = OfflineRecognizerConfig(
                        featConfig = FeatureConfig(dither = family.featureDither),
                        modelConfig = modelConfig,
                        decodingMethod = decodingMethod,
                        maxActivePaths = quality.sherpaMaxActivePaths,
                    ),
                ),
                // Only the families that can honour a target are translating,
                // whatever the caller asked for.
                translating = family.acceptsLanguage && translateTo.isNotEmpty(),
            )
        }
    }
}

/**
 * A decoded chunk, and the language the model said it was.
 *
 * Only SenseVoice fills the language in — it decodes a `<|en|>`-style tag as its
 * first token. The other families leave it empty, and the writing styles then
 * fall back to inspecting the text, exactly as they always have.
 */
internal data class SherpaTranscript(val text: String, val language: String = "") {

    /** The first language anything reported wins; later chunks rarely disagree. */
    fun append(next: SherpaTranscript, deduplicateOverlap: Boolean): SherpaTranscript =
        SherpaTranscript(
            text = SherpaTranscriptMerger.append(text, next.text, deduplicateOverlap),
            language = language.ifEmpty { next.language },
        )

    companion object {
        val EMPTY = SherpaTranscript("")

        /**
         * Turns SenseVoice's `<|en|>` token into `en`.
         *
         * Anything that does not look like a language code becomes empty rather
         * than being passed on: the first token is a language tag by convention
         * and not by guarantee, and a bogus code would pick the wrong
         * punctuation with more confidence than no code at all.
         */
        fun languageCode(raw: String?): String {
            val trimmed = raw?.trim()?.removeSurrounding("<|", "|>")?.trim().orEmpty()
            return trimmed.takeIf { it.length in 2..3 && it.all(Char::isLetter) }
                ?.lowercase()
                .orEmpty()
        }
    }
}

/**
 * Some attention-based models occasionally return no tokens for a longer
 * waveform even though shorter speech from the same recording is recognized.
 * Retry only that empty chunk as two overlapping smaller streams; successful
 * chunks never pay the extra inference cost.
 */
internal object SherpaEmptyChunkRecovery {

    /**
     * @param deduplicateOverlap false when the caller is translating. The two
     *   halves overlap so the word crossing the centre survives, and matching
     *   the repeat back out only works when the same audio returns the same
     *   words — which is exactly what a translator does not promise.
     * @param recoverAudibleShortInput true when an empty answer here cannot be
     *   the ordinary short trailing overlap. See [SherpaRecognizer.transcribe]
     *   for how that is decided.
     */
    fun decode(
        samples: FloatArray,
        decodeOnce: (FloatArray) -> SherpaTranscript,
        deduplicateOverlap: Boolean = true,
        recoverAudibleShortInput: Boolean = false,
    ): SherpaTranscript {
        // Room tone is the ordinary reason for an empty answer, and the cheapest
        // thing to rule out -- before the first decode, not after it. The scan
        // costs one pass over the samples; a decode of a pause costs inference
        // to be told what the scan already knew. iOS guards in the same place.
        if (SherpaLongAudio.isEffectivelySilent(samples)) return SherpaTranscript.EMPTY

        val firstAttempt = decodeOnce(samples)
        if (firstAttempt.text.isNotEmpty()) return firstAttempt

        // Length is what the split recovers from, so a window that is not long
        // enough to be dropped for its length has nothing there to recover --
        // unless the caller has already established that this window is the
        // whole of what the user has said so far.
        val short =
            samples.size <= SherpaLongAudio.MIN_SUSPECT_CHUNK_SECONDS * SherpaLongAudio.SAMPLE_RATE
        if (!recoverAudibleShortInput && short) return firstAttempt

        // The two extra rungs are for short speech and stay there. A ten-second
        // window is already the length the split recovers from, and padding it
        // by half a second either side buys nothing for the cost of a whole
        // further decode -- so a long window keeps the established ladder and
        // its predictable latency however little has been decoded before it.
        if (short) {
            // A fresh offline stream can recover a nondeterministic no-token
            // answer.
            val repeated = decodeOnce(samples)
            if (repeated.text.isNotEmpty()) return repeated
            // Very short speech can begin or end too close to the encoder
            // context. Half a second either side moves it inside, and is
            // bounded so a recording cannot grow its own decode cost.
            val padding = FloatArray(SherpaLongAudio.SAMPLE_RATE / 2)
            val padded = decodeOnce(padding + samples + padding)
            if (padded.text.isNotEmpty()) return padded
        }

        // Retain context on both sides of the recovery boundary. A plain
        // midpoint split could rescue an empty window while still deleting the
        // word crossing its exact centre. One split and no more: a half-length
        // window that is still empty is not being lost to its length, and
        // subdividing again multiplies the decodes for nothing.
        // Clamped to a quarter of the waveform: without it a recording shorter
        // than the overlap itself hands both halves the entire thing, which is
        // two identical full decodes and no split at all. Only reachable now
        // that a short complete recording gets this far.
        val midpoint = samples.size / 2
        val halfOverlap = (SherpaLongAudio.OVERLAP_MILLIS * SherpaLongAudio.SAMPLE_RATE / 2_000)
            .coerceAtMost(midpoint / 2)
        val leftEnd = (midpoint + halfOverlap).coerceAtMost(samples.size)
        val rightStart = (midpoint - halfOverlap).coerceAtLeast(0)
        return decodeOnce(samples.copyOfRange(0, leftEnd))
            .append(
                decodeOnce(samples.copyOfRange(rightStart, samples.size)),
                deduplicateOverlap = deduplicateOverlap,
            )
    }
}

File: ios/VocaPhoneApp/Models/SherpaRecognizer.swift
import Foundation

/// One sherpa-onnx recognizer. The C bridge keeps the Swift side independent
/// of the large generated API surface and exposes only the offline ASR calls
/// VocaPhone needs.
final class SherpaRecognizer: @unchecked Sendable {
    private let native: UnsafeMutableRawPointer
    /// sherpa-onnx recognizers are mutable native objects. Incremental decoding
    /// runs off the main actor while a retry or settings preparation can also
    /// reach this instance, so one recognizer must never decode two streams at
    /// once.
    private let decodeLock = NSLock()
    /// Whether this recognizer was built to translate, which changes how a
    /// recording too long for one decode is put back together. See `transcribe`.
    private let translating: Bool
    /// The feature noise this family needs, applied to the waveform because the
    /// pinned runtime has no `dither` field to ask for it. See
    /// `SherpaFeatureDither` and `SherpaFamily.featureDither`.
    private let featureDither: Float

    private init(
        native: UnsafeMutableRawPointer,
        translating: Bool = false,
        featureDither: Float = 0
    ) {
        self.native = native
        self.translating = translating
        self.featureDither = featureDither
    }

    deinit {
        VocaPhoneSherpaDestroy(native)
    }

    static func create(
        model: LocalModelDescriptor,
        directory: URL,
        language: String,
        threads: Int,
        quality: TranscriptionQuality,
        translateTo: String = ""
    ) throws -> SherpaRecognizer {
        guard let family = model.sherpaFamily else {
            throw LocalModelManagerError.unsupportedModel(model.id)
        }

        func path(_ name: String) throws -> String {
            let url = directory.appendingPathComponent(name)
            guard FileManager.default.fileExists(atPath: url.path) else {
                throw LocalModelManagerError.integrityFileMissing(name)
            }
            return url.path
        }

        /// The quantized file where the model ships one, the plain file where
        /// it does not. Upstream quantizes per graph rather than per model:
        /// GigaAM's transducer ships an int8 encoder beside a full-precision
        /// decoder and joiner, because those two are small enough that
        /// quantizing them costs accuracy for nothing.
        func quantizedOrPlain(_ stem: String) throws -> String {
            let int8 = directory.appendingPathComponent("\(stem).int8.onnx")
            return FileManager.default.fileExists(atPath: int8.path)
                ? int8.path
                : try path("\(stem).onnx")
        }

        let tokens = try path("tokens.txt")
        let models: [String]
        switch family {
        case .nemoTransducer:
            models = try [
                quantizedOrPlain("encoder"), quantizedOrPlain("decoder"),
                quantizedOrPlain("joiner"), ""
            ]
        case .moonshine:
            models = try [
                path("preprocess.onnx"), path("encode.int8.onnx"),
                path("uncached_decode.int8.onnx"), path("cached_decode.int8.onnx")
            ]
        case .moonshineV2:
            models = try [path("encoder_model.ort"), path("decoder_model_merged.ort"), "", ""]
        case .canary:
            models = try [path("encoder.int8.onnx"), path("decoder.int8.onnx"), "", ""]
        case .senseVoice, .dolphinCtc, .paraformer:
            models = try [path("model.int8.onnx"), "", "", ""]
        case .nemoCtc:
            let name = FileManager.default.fileExists(
                atPath: directory.appendingPathComponent("model.onnx").path
            ) ? "model.onnx" : "model.int8.onnx"
            models = try [path(name), "", "", ""]
        }

        let languageHint: String
        switch family {
        case .canary:
            // Canary's config has no detection mode, so "auto" has to become a
            // real code — English is the safest guess and the one upstream's
            // own examples use.
            languageHint = language == "auto" ? "en" : language
        case .senseVoice:
            languageHint = language == "auto" ? "" : language
        default:
            languageHint = ""
        }
        // Only Canary has a target at all; the bridge falls back to the source
        // when this is empty, which is what transcription is.
        let targetHint = family == .canary ? translateTo : ""

        // Never `quality.sherpaDecodingMethod` on its own: a family that does
        // not support beam search answers it by killing the process.
        let decodingMethod = family.decodingMethod(for: quality)
        let native: UnsafeMutableRawPointer? = models[0].withCString { model1 in
            models[1].withCString { model2 in
                models[2].withCString { model3 in
                    models[3].withCString { model4 in
                        tokens.withCString { tokensPointer in
                            languageHint.withCString { languagePointer in
                                targetHint.withCString { targetPointer in
                                    decodingMethod.withCString { decodingMethodPointer in
                                        VocaPhoneSherpaCreate(
                                            Int32(family.bridgeValue),
                                            model1, model2, model3, model4,
                                            tokensPointer, languagePointer, targetPointer,
                                            Int32(threads), decodingMethodPointer,
                                            quality.sherpaMaxActivePaths
                                        )
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        guard let native else {
            throw LocalModelManagerError.engineLoadFailed(model.id)
        }
        // Only the families that can honour a target are translating, whatever
        // the caller asked for.
        return SherpaRecognizer(
            native: native,
            translating: family.acceptsLanguage && !translateTo.isEmpty,
            featureDither: family.featureDither
        )
    }

    /// Decodes the whole recording, in windows if it is longer than one decode
    /// should be.
    ///
    /// Those windows are where translation and transcription part company. A
    /// transcriber returns the same words for the same audio, so a window that
    /// retains a little of the one before it can have the repeat matched and
    /// removed. A translator does not: the retained audio is translated again
    /// inside a different sentence, comes back in different words, and nothing
    /// can pair the two. Worse, a merger still looking for a repeat can only
    /// match a phrase the speaker genuinely said twice — and delete it.
    ///
    /// A boundary the silence search found is already cut clean here, so only
    /// a guessed one carries audio across, and while translating that seam is
    /// left as it decoded rather than merged by matching words.
    func transcribe(_ samples: [Float]) -> SherpaDecodeOutcome {
        var transcript = SherpaTranscript.empty
        var previousEnd = 0
        var loudestSoFar = 0.0
        for chunk in SherpaLongAudio.chunks(samples) {
            let bounded = Array(samples[chunk.start..<chunk.endExclusive])
            // Whether a short empty answer is ordinary is a question about the
            // audio this window did not inherit from the one before it. A final
            // window that is only the retained overlap has nothing new in it,
            // and an empty answer there is correct. A final window carrying a
            // whole further sentence is the reported failure, and it is that
            // whether or not an earlier window already produced text — judging
            // it by the transcript so far is what let a closing sentence
            // disappear behind a successful opening one.
            let newRegion = SherpaLongAudio.newRegion(
                of: samples, chunk: chunk, previousEnd: previousEnd
            )
            let newRegionLevel = SherpaLongAudio.loudestFrame(newRegion)
            let carriesNewSpeech = SherpaLongAudio.carriesRecoverableSpeech(
                newRegion: newRegion,
                inheritsAudio: chunk.overlapsPrevious,
                loudestFrame: newRegionLevel,
                loudestFrameSoFar: loudestSoFar
            )
            let outcome = SherpaEmptyChunkRecovery.decode(
                samples: bounded,
                decodeOnce: decode,
                deduplicateOverlap: !translating,
                recoverAudibleShortInput: carriesNewSpeech
            )
            guard case let .decoded(decoded) = outcome else { return outcome }
            loudestSoFar = max(loudestSoFar, newRegionLevel)
            previousEnd = chunk.endExclusive
            transcript = transcript.appending(
                decoded, deduplicateOverlap: chunk.overlapsPrevious && !translating
            )
        }
        return .decoded(
            SherpaTranscript(
                text: transcript.text.trimmingCharacters(in: .whitespacesAndNewlines),
                language: transcript.language
            )
        )
    }

    func transcribeChunk(_ samples: [Float]) -> SherpaDecodeOutcome {
        SherpaEmptyChunkRecovery.decode(
            samples: samples, decodeOnce: decode, deduplicateOverlap: !translating
        )
    }

    private func decode(_ samples: [Float]) -> SherpaDecodeOutcome {
        guard !samples.isEmpty else { return .failed(.invalidArgument) }
        let samples = SherpaFeatureDither.applied(to: samples, dither: featureDither)
        decodeLock.lock()
        defer { decodeLock.unlock() }
        var output = [CChar](repeating: 0, count: 131_072)
        // Only ever holds a short tag such as "<|en|>".
        var languageOutput = [CChar](repeating: 0, count: 32)
        let result = samples.withUnsafeBufferPointer { sampleBuffer in
            output.withUnsafeMutableBufferPointer { outputBuffer in
                languageOutput.withUnsafeMutableBufferPointer { languageBuffer in
                    VocaPhoneSherpaDecode(
                        native,
                        sampleBuffer.baseAddress,
                        Int32(samples.count),
                        outputBuffer.baseAddress,
                        Int32(outputBuffer.count),
                        languageBuffer.baseAddress,
                        Int32(languageBuffer.count)
                    )
                }
            }
        }
        // A negative status is the engine failing to answer, and it must never
        // reach the caller looking like a model that answered nothing.
        guard result >= 0 else { return .failed(.forStatus(result)) }
        return .decoded(
            SherpaTranscript(
                text: Self.string(from: output),
                language: SherpaTranscript.languageCode(Self.string(from: languageOutput))
            )
        )
    }

    private static func string(from buffer: [CChar]) -> String {
        String(decoding: buffer.prefix { $0 != 0 }.map { UInt8(bitPattern: $0) }, as: UTF8.self)
    }
}

private extension SherpaFamily {
    var bridgeValue: Int {
        switch self {
        case .nemoTransducer: 0
        case .senseVoice: 1
        case .moonshine: 2
        case .dolphinCtc: 3
        case .canary: 4
        case .nemoCtc: 5
        case .paraformer: 6
        case .moonshineV2: 7
        }
    }
}

extension SherpaIncrementalSession {
    /// The session lives in the shared target and takes a decode closure so it
    /// can be tested without the native engine; this is the one call site that
    /// has a real recognizer to give it.
    convenience init(chunks: AsyncStream<Data>, recognizer: SherpaRecognizer) {
        self.init(chunks: chunks) { recognizer.transcribeChunk($0) }
    }
}
